### PR TITLE
Show "cast down" animation for heroes on the battlefield

### LINKED ISSUE #7450: Battle screen, missing angled hero animation
### Preliminary checks

- [X] I've checked that there aren't [**other open issues**](https://github.com/ihhub/fheroes2/issues?q=is%3Aissue) on the same topic.
- [X] I've checked that this issue is reproducible on the [**latest snapshot build**](https://ihhub.github.io/fheroes2/INSTALL.html#snapshots-latest-builds).

### Platform

Windows

### Describe the bug

Fheroes2:
image

OG:
image

In the original game while casting on creatures standing in the lower part of the screen hero rises scepter a bit lower, than normal cast.

While casting on creatures standing on the cells (and to the left of them) hero should rise scepter lower:
angle attack

### Save file

None.

### Additional info

_No response_

### PR DESCRIPTION
**Pull request by Districh-ru**
fix #7450 
Show "cast down" animation for heroes on the battlefield if the target is lower than 3rd row

изображение

### CODE REVIEW

File: src/fheroes2/battle/battle_interface.cpp [hunk 1]
```diff
@@ -4220,7 +4220,7 @@ void Battle::Interface::RedrawActionSpellCastPart1( const Spell & spell, int32_t
     if ( caster ) {
         OpponentSprite * opponent = caster->GetColor() == arena.GetArmy1Color() ? _opponent1.get() : _opponent2.get();
         if ( opponent ) {
-            opponent->SetAnimation( spell.isApplyWithoutFocusObject() ? OP_CAST_MASS : OP_CAST_UP );
+            opponent->SetAnimation( spell.isApplyWithoutFocusObject() ? OP_CAST_MASS : ( dst > 32 ? OP_CAST_DOWN : OP_CAST_UP ) );
```
Comment (ihhub) [inline src/fheroes2/battle/battle_interface.cpp:4223]: Please avoid using a ternary operator inside another ternary operator. It is better to have if-else condition for readability purposes.

Comment (Districh-ru) [reply src/fheroes2/battle/battle_interface.cpp]: Sure)

**Comment by Branikolog** [comment]
Hi, @Districh-ru !
Not all casts on creatures on a forth row and lower should be with "cast down" hero animation.
OG:
image

In #7450 I've attached a screenshot, mentioning the cells, that should be the case for "cast down" animation.
angle attack 2

**Comment by Districh-ru** [comment]
Hi, @Branikolog, I looked at that picture, read a text, but for some reason decided to divide screen horizontally, not diagonally for cast down animation.
I'll change the logic soon.

**Comment by Districh-ru** [comment]
@Branikolog, I've fixed the logic to correspond your screenshot. For the right hero I used the same but horizontally mirrored logic: after the 3rd raw the number of columns is increased by one each row from the side of the hero.
Could you please check the solution in your spare time.

**Review by Branikolog — APPROVED** [review]
Now everything looks good!
Thanks for adding new animation, @Districh-ru !

File: src/fheroes2/battle/battle_interface.cpp [hunk 2]
```diff
@@ -4216,11 +4216,26 @@ void Battle::Interface::RedrawActionSpellCastPart1( const Spell & spell, int32_t
...
+        const bool isLeftOpponent = caster->GetColor() == arena.GetArmy1Color();
+        opponent = isLeftOpponent ? _opponent1.get() : _opponent2.get();
+        if ( opponent != nullptr ) {
+            if ( isMassSpell ) {
+                opponent->SetAnimation( OP_CAST_MASS );
+            }
+            else {
+                // The cast down is applied below the 2rd battlefield row (count is started from 0)
+                // and for the (rowNumber - 2) columns starting from the side of the hero.
+                isCastDown = isLeftOpponent ? ( dst % 11 < dst / 11 - 2 ) : ( 10 - dst % 11 < dst / 11 - 2 );
```
Comment (ihhub) [inline src/fheroes2/battle/battle_interface.cpp:4234]: Please add parenthesis for **dst % 11** in both cases.

**Review by ihhub — COMMENTED** [review]
Hi @Districh-ru , I put one comment for code readability. Would you mind to take a look?

**Review by ihhub — APPROVED** [review]

**Comment by ihhub** [comment]
@Districh-ru , thank you so much for this fix!

**Pull request merged**